swig/ctc_beam_search_decoder.cpp

```diff
@@ -130,9 +148,39 @@ std::vector<std::pair<double, std::vector<int>>> ctc_beam_search_decoder(
             float score = 0.0;
             std::vector<std::string> ngram;
             ngram = ext_scorer->make_ngram(prefix_to_score);
-            score = ext_scorer->get_log_cond_prob(ngram) * ext_scorer->alpha;
+
+            // hot words boosting
+            float hot_boost = 0.0;
+            if (hotwords_scorer != nullptr && !hotwords_scorer->hotwords_dict.empty()){
+                    std::unordered_map<std::string, float>::const_iterator iter;
+                    for (size_t index=0; index < ngram.size(); index ++ ) {
+                        std::string word = "";
+                        // character-based language model, combining chinese characters into words
+                        if (ext_scorer->is_character_based()) {
```

Comment:
I am not fully understand why we need a external arpa file to launch the hotword scorer when we set use_ngram_score=False. If the ext_scorer->is_character_based() is the reason, how about setting an independent flag?

Reply:
I see. So, if we use a 4-gram-arpa as ext_scorer, we can't handle hot words more than 4 like this one https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/blob/main/models/hotwords.yaml#L4, right? 

What do you think if we copy that make_ngram into HotWord Class? Such that we could set the hot word ngram max_length as well as is_character_or_not.

Since it would be a little bit confusing, when we set the lm_path, but our final decoding results didn't use ngram lm.

Reply:
> 

1. If 4-gram-arpa as ext_scorer，only support a maximum length of 4 Chinese characters.
2. I plan to copy make_ngram into HotWordsBoosting Class, and exposing the length parameter of hotwords.

Reply:
In the latest commit, we separated hotwords scoring from language model scoring. The latest hotwords results on speech_asr_aishell1_hotwords_testsets and AISHELL-1 Test dataset can be found [here](https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords).